### PR TITLE
[Workers] Add tutorial: Live cursors & RPC with Next.js and Durable Objects

### PR DESCRIPTION
**Pull request by exectx**
### Summary

Hi, here's the tutorial for the Cloudflare Developer Spotlight program. It's covers using Next.js Workers with Durable Objects and Workers RPC to visualize connected users cursors via WebSockets. I've also shortened the original title a bit

### Documentation checklist
- [x] The [documentation style guide](https://developers.cloudflare.com/style-guide/) has been adhered to.

### CODE REVIEW

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 1]
```diff
@@ -0,0 +1,764 @@
...
+	<PackageManagers type="add" pkg="nanoid" frame="none" />
+
+3. Define workspaces for each Worker:
+
+	<Tabs> <TabItem label="npm">
+
+	Update your `package.json` file.
+
+
+	```json title="package.json" ins={14-17}
```
Comment (bllchmbrs) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx:76]: Hyperlint dev here, just to make sure I understand, this indentation is significant, right? It's a part of the step by step workflow and therefore indented?

Comment (exectx) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx:76]: You’re right, it actually fails to build if the indentation is removed. There’s something about MDX here that might be causing it. I tried using spaces, but it looks like tabs are required

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 2]
```diff
@@ -0,0 +1,764 @@
...
+- An ID is generated for each user that navigates to the application, which is used for identifying the WebSocket connection in the Durable Object.
+- Once the WebSocket connection is established, the application sends a message to the WebSocket Durable Object to determine the current number of connected users.
+- A user can close all active WebSocket connections via a Next.js server action that uses an RPC method.
+- It handles WebSocket and mouse movement events to update the location of other users' cursors in the UI and to send updates about the user's own cursor, as well as join and leave WebSocket events.
+
+![Animated gif of real-time Next.js app for visualizing live cursors](~/assets/images/workers/tutorials/live-cursors-nextjs/demo-live-cursors-nextjs-do.gif)
+
+---
+
+## Create a Next.js Workers Project
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
## 1. Create a Next.js Workers Project
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 3]
```diff
@@ -0,0 +1,764 @@
...
+	```yaml title="pnpm-workspace.yaml"
+	packages:
+		- "worker"
+		- "."
+	```
+	</TabItem> </Tabs>
+
+</Steps>
+
+## Create a Durable Object Worker
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
## 2. Create a Durable Object Worker
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 4]
```diff
@@ -0,0 +1,764 @@
...
+		product="workers"
+		params={{
+			category: "hello-world",
+			type: "Hello World Worker using Durable Objects",
+			lang: "JavaScript / TypeScript",
+		}}
+	/>
+</Steps>
+
+## Setup Workers
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 5]
```diff
@@ -0,0 +1,764 @@
...
+
+	[[durable_objects.bindings]]
+	name = "CURSOR_SESSIONS"
+	class_name = "CursorSessions"
+
+	[[migrations]]
+	tag = "v1"
+	new_classes = ["CursorSessions"]
+	```
+1. Initialize Durable Object class:
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
2. Initialize the Durable Object class:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 6]
```diff
@@ -0,0 +1,764 @@
...
+	```
+
+	Now update `worker-configuration.d.ts` by running:
+
+	```sh
+	cd worker && npm run cf-typegen
+	```
+
+
+1. Update the Durable Object to manage WebSockets:
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
3. Update the Durable Object to manage WebSockets:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx (hunk 5 above)
Comment (db-cloudflare) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: It would also be worth specifying the file being edited here.

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 7]
```diff
@@ -0,0 +1,764 @@
...
+			category: "hello-world",
+			type: "Hello World Worker using Durable Objects",
+			lang: "JavaScript / TypeScript",
+		}}
+	/>
+</Steps>
+
+## Setup Workers
+
+### Worker project
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
## 2. Build Durable Object Worker functionality
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 8]
```diff
@@ -0,0 +1,764 @@
...
+	<PackageManagers type="run" pkg="dev" frame="none" />
+
+1. Add the Worker host as a Next.js public environment variable:
+	```text title=".env.local" ins={1}
+	NEXT_PUBLIC_WS_HOST=localhost:8787
+	```
+
+</Steps>
+
+### Next.js project
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
## 3. Build Next.js Worker functionality
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 9]
```diff
@@ -0,0 +1,764 @@
...
+	Build the app:
+	<PackageManagers type="run" pkg="build:worker" frame="none" />
+
+	Now, you can run Next.js development server:
+	<PackageManagers type="run" pkg="dev" frame="none" />
+
+</Steps>
+
+
+## Deploy the project
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
## 4. Deploy the project
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 10]
```diff
@@ -0,0 +1,764 @@
...
+					"Content-Type": "text/plain",
+				},
+			});
+		},
+	} satisfies ExportedHandler<Env>;
+
+	```
+
+
+1. Extend the `WorkerEntrypoint` class for RPC:
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
4. Extend the `WorkerEntrypoint` class for RPC:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 11]
```diff
@@ -0,0 +1,764 @@
...
+		}
+	}
+
+	export default {
+		async fetch(request, env, ctx) {
+			if (request.url.match("/ws")) {
+	// ...
+	```
+
+1. Run the Worker locally:
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
5. Run the Worker locally:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 12]
```diff
@@ -0,0 +1,764 @@
...
+	<PackageManagers type="run" pkg="dev" frame="none" />
+
+1. Add the Worker host as a Next.js public environment variable:
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
6. Add the Worker host as a Next.js public environment variable:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 13]
```diff
@@ -0,0 +1,764 @@
...
+	```ts title="env.d.ts"
+	interface CloudflareEnv {
+		CURSOR_SESSIONS: DurableObjectNamespace<
+			import("./worker/src/index").CursorSessions
+		>;
+		RPC_SERVICE: Service<import("./worker/src/index").SessionsRPC>;
+		ASSETS: Fetcher;
+	}
+	```
+1. Create server actions that make use of RPC methods:
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
2. Create server actions that make use of RPC methods:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 14]
```diff
@@ -0,0 +1,764 @@
...
+					<div className="px-2 space-y-2">
+						<Cursors id={id}></Cursors>
+					</div>
+				</div>
+			</main>
+		);
+	}
+	```
+
+1. Create a client component to manage WebSocket and mouse movement events:
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
3. Create a client component to manage WebSocket and mouse movement events:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 15]
```diff
@@ -0,0 +1,764 @@
...
+	</Details>
+	- It starts the WebSocket connection.
+	- It handles 4 types of WebSocket messages:
+		- `join`. Received when a new WebSocket connection is established.
+		- `quit`. Received when a WebSocket connection is closed.
+		- `move`. Received when a user's cursor moves.
+		- `get-cursors-response`. Received	when a client sends a `get-cursors` message, which is triggered once the WebSocket connection is open.
+	- It sends the user's cursor coordinates to the WebSocket server, which broadcasts them to all active WebSocket clients.
+
+1. Run Next.js locally:
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
4. Run Next.js locally:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 16]
```diff
@@ -0,0 +1,764 @@
...
+	<PackageManagers type="run" pkg="deploy" frame="none" />
+
+	A URL is generated for the Worker, which will be used to update Next.js `NEXT_PUBLIC_WS_HOST` environment variable in your `.env.local` file.
+
+	```txt title=".env.local" ins={2} del={1}
+	NEXT_PUBLIC_WS_HOST=localhost:8787
+	NEXT_PUBLIC_WS_HOST=<DEPLOYED-WORKER-HOST>
+	```
+
+1. Change into your root directory and deploy your Next.js app:
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
2. Change into your root directory and deploy your Next.js app:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 17]
```diff
@@ -0,0 +1,764 @@
...
+		}}
+	/>
+
+2. Change into your new directory:
+
+	```sh
+	cd next-rpc
+	```
+
+	Install [nanoid](https://www.npmjs.com/package/nanoid):
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
3. Install [nanoid](https://www.npmjs.com/package/nanoid) so that string IDs can be generated for clients:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 18]
```diff
@@ -0,0 +1,764 @@
...
+	<PackageManagers type="add" pkg="nanoid" frame="none" />
+
+3. Define workspaces for each Worker:
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
4. Define workspaces for each Worker:
```

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx (hunk 17 above)
Comment (db-cloudflare) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: If the added description on how nanoid is used isn't accurate feel free to change the copy.

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 19]
```diff
@@ -0,0 +1,764 @@
...
+		CURSOR_SESSIONS: DurableObjectNamespace<
+			import("./worker/src/index").CursorSessions
+		>;
+		RPC_SERVICE: Service<import("./worker/src/index").SessionsRPC>;
+		ASSETS: Fetcher;
+	}
+	```
+1. Create server actions that make use of RPC methods:
+
+	The Service binding `RPC_SERVICE` is used instead of invoking Durable Object RPC methods because of the limitation mentioned in the note above.
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: I'd suggest adding some copy to describe how nanoid is used in the code block as well.

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 20]
```diff
@@ -0,0 +1,764 @@
+---
+updated: 2024-11-05
+difficulty: Intermediate
+content_type: 📝 Tutorial
+pcx_content_type: tutorial
+title: Live cursors & RPC with Next.js and Durable Objects
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
title: Display live cursors with RPC, Next.js and Durable Objects
```

Comment (db-cloudflare) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: I suggest making the title a present tense action to match our other tutorials. Maybe something like my suggested example, but feel free to change it.

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 21]
```diff
@@ -0,0 +1,764 @@
...
+	{
+		"name": "next-rpc",
+		"version": "0.1.0",
+		"private": true,
+		"scripts": {
+			"dev": "next dev",
+			"build": "next build",
+			"start": "next start",
+			"lint": "next lint",
+			"deploy": "cloudflare && wrangler deploy",
```
[the remark this answers is not in the document]
Comment (exectx) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx:86]: The `cloudflare` command comes from the open nextjs package, I can't change it

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 22]
```diff
@@ -0,0 +1,764 @@
...
+			"preview": "cloudflare && wrangler dev",
+			"cf-typegen": "wrangler types --env-interface CloudflareEnv env.d.ts"
+		},
+		"workspaces": [
+			".",
+			"worker"
```
[the remark this answers is not in the document]
Comment (exectx) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx:92]: In this case `"worker"` is the name of the directory for the workspace, so I can’t change it.

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 23]
```diff
@@ -0,0 +1,764 @@
...
+	name = "CURSOR_SESSIONS"
+	class_name = "CursorSessions"
+
+	[[migrations]]
+	tag = "v1"
+	new_classes = ["CursorSessions"]
+	```
+1. Initialize Durable Object class:
+	```ts title="worker/src/index.ts"
+	import { DurableObject } from 'cloudflare:workers';
```
[the remark this answers is not in the document]
Comment (exectx) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx:156]: The indentation is necessary to prevent MDX from crashing and is part of the step-by-step workflow.
`'cloudflare:workers'` is a Cloudflare module, it is case-sensitive and cannot be changed.

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 24]
```diff
@@ -0,0 +1,761 @@
...
+tags:
+  - Next.js
+  - WebSockets
+  - RPC
+languages:
+  - TypeScript
+---
+import { Render, PackageManagers, Steps, TabItem, Tabs, Details } from "~/components";
+
+In this tutorial, you will learn how to build a real-time [Next.js](https://nextjs.org/) app that displays the live cursor location of each connected user using [Durable Objects](/durable-objects/), Workers' built-in [RPC (Remote Procedure Call)](/workers/runtime-apis/rpc/) system and the [OpenNext](https://opennext.js.org/cloudflare) Cloudflare adapter.
```
Comment (ToriLindsay) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
In this tutorial, you will learn how to build a real-time [Next.js](https://nextjs.org/) app that displays the live cursor location of each connected user using [Durable Objects](/durable-objects/), the Workers' built-in [RPC (Remote Procedure Call)](/workers/runtime-apis/rpc/) system, and the [OpenNext](https://opennext.js.org/cloudflare) Cloudflare adapter.
```
For clarity

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 25]
```diff
@@ -0,0 +1,761 @@
...
+		- "worker"
+		- "."
+	```
+	</TabItem> </Tabs>
+
+</Steps>
+
+## 2. Create a Durable Object Worker
+
+This Worker will manage the Durable Object and also have internal API's
```
Comment (ToriLindsay) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
This Worker will manage the Durable Object and also have internal APIs
```
Grammar fix

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 26]
```diff
@@ -0,0 +1,761 @@
...
+
+	Now update `worker-configuration.d.ts` by running:
+
+	```sh
+	cd worker && npm run cf-typegen
+	```
+
+
+3. Update the Durable Object to manage WebSockets:
+	- The main `fetch` handler routes requests with a `/ws` URL to the `CursorSessions` Durable Object where a WebSocket connection is stablished.
```
Comment (ToriLindsay) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
	- The main `fetch` handler routes requests with a `/ws` URL to the `CursorSessions` Durable Object where a WebSocket connection is established.
```
Typo

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 27]
```diff
@@ -0,0 +1,761 @@
...
+	- The `CursorSessions` class manages WebSocket connections, session states, and broadcasts messages to other connected clients.
+		- When a new WebSocket connection is established, the Durable Object broadcasts a `join` message to all connected clients; similaryly, a `quit` message is broadcasted when a client disconnects.
```
Comment (ToriLindsay) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]:
```suggestion
		- When a new WebSocket connection is established, the Durable Object broadcasts a `join` message to all connected clients; similarly, a `quit` message is broadcast when a client disconnects.
```
Typo / Grammar

**Review by ToriLindsay — APPROVED** [review]
Fixing typos and grammar

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 28]
```diff
@@ -0,0 +1,761 @@
...
+		"name": "next-rpc",
+		"version": "0.1.0",
+		"private": true,
+		"scripts": {
+			"dev": "next dev",
+			"build": "next build",
+			"start": "next start",
+			"lint": "next lint",
+			"deploy": "cloudflare && wrangler deploy",
+"preview": "Cloudflare && wrangler dev",
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: Shouldn't it be `cloudflare` and not `Cloudflare`?

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 29]
```diff
@@ -0,0 +1,761 @@
...
+
+## 1. Create a Next.js Workers Project
+
+<Steps>
+
+1. Run the following command to create your Next.js Worker named `next-rpc`:
+
+	<PackageManagers
+		type="create"
+		pkg="cloudflare@latest next-rpc"
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: We should move `next-rpc` to the `args` parameter

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 30]
```diff
@@ -0,0 +1,761 @@
...
+		args={"worker"}
+	/>
+
+	<Render
+		file="c3-post-run-steps"
+		product="workers"
+		params={{
+			category: "hello-world",
+			type: "Hello World Worker using Durable Objects",
+			lang: "JavaScript / TypeScript",
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: Since we are using TypeScript, let's specify that here. This should be updated to `lang: "TypeScript"`

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 31]
```diff
@@ -0,0 +1,761 @@
...
+
+	[[durable_objects.bindings]]
+	name = "CURSOR_SESSIONS"
+	class_name = "CursorSessions"
+
+	[[migrations]]
+	tag = "v1"
+	new_classes = ["CursorSessions"]
+	```
+2. Initialize the Durable Object class in your `worker/src/index.ts` file:
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: We should add a few lines here explaining the intention of the code below

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 32]
```diff
@@ -0,0 +1,761 @@
...
+	```
+
+
+3. Update the Durable Object to manage WebSockets:
+	- The main `fetch` handler routes requests with a `/ws` URL to the `CursorSessions` Durable Object where a WebSocket connection is established.
+	- The `CursorSessions` class manages WebSocket connections, session states, and broadcasts messages to other connected clients.
+		- When a new WebSocket connection is established, the Durable Object broadcasts a `join` message to all connected clients; similarly, a `quit` message is broadcast when a client disconnects.
+		- It tracks each WebSocket client's last cursor position under the `move` message, which is broadcasted to all active clients.
+		- When a `get-cursors` message is received, it sends the number of currently active clients to the specific client that requested it.
+
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: We should modify the sentences here. It sounds like we have already implemented this in the code. Either move it after the code block, or reword it to let the user know this is what the below code block implements

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 33]
```diff
@@ -0,0 +1,761 @@
...
+			super(ctx, env);
+			this.sessions = new Map();
+			this.ctx.getWebSockets().forEach((ws) => {
+				const meta = ws.deserializeAttachment();
+				this.sessions.set(ws, { ...meta });
+			});
+		}
+
+		broadcast(message: WsMessage, self?: string) {
+			this.sessions.forEach((session, ws) => {
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: Since you are using the WebSocket Hibernation API, we should use the [`getWebSockets` method](https://developers.cloudflare.com/durable-objects/api/state/#getwebsockets)

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 34]
```diff
@@ -0,0 +1,761 @@
...
+		}
+	}
+
+	export default {
+		async fetch(request, env, ctx) {
+			if (request.url.match("/ws")) {
+	// ...
+	```
+
+5. Run the Worker locally:
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: - Should this be run from the worker directory or the root (next.js) directory?
- What should be expected here? After starting the development server, should I go to localhost:8787?

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 35]
```diff
@@ -0,0 +1,761 @@
...
+6. Add the Worker host as a Next.js public environment variable:
+	```text title=".env.local" ins={1}
+	NEXT_PUBLIC_WS_HOST=localhost:8787
+	```
+
+</Steps>
+
+## 4. Build Next.js Worker functionality
+<Steps>
+1. In your `wrangler.toml` file, declare the external Durable Object binding and the Service binding to `SessionsRPC`:
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: Should it be the `next-rpc/wrangler.toml` file or the worker's?

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 36]
```diff
@@ -0,0 +1,761 @@
...
+	```ts title="worker/src/index.ts"
+	import { DurableObject } from 'cloudflare:workers';
+
+	export type WsMessage =
+		| { type: "message"; data: string }
+		| { type: "quit"; id: string }
+		| { type: "join"; id: string }
+		| { type: "move"; id: string; x: number; y: number }
+		| { type: "get-cursors" }
+		| { type: "get-cursors-response"; sessions: Session[] };
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx:164]: `Session` is never declared. VS Code shows an error whenever the `Session` type is being referred

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 37]
```diff
@@ -0,0 +1,761 @@
...
+	}
+	```
+	</Details>
+	- It starts the WebSocket connection.
+	- It handles 4 types of WebSocket messages:
+		- `join`. Received when a new WebSocket connection is established.
+		- `quit`. Received when a WebSocket connection is closed.
+		- `move`. Received when a user's cursor moves.
+		- `get-cursors-response`. Received	when a client sends a `get-cursors` message, which is triggered once the WebSocket connection is open.
+	- It sends the user's cursor coordinates to the WebSocket server, which broadcasts them to all active WebSocket clients.
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: While the overview is good, a detailed explanation would be really helpful here.

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 38]
```diff
@@ -0,0 +1,761 @@
...
+			"build:worker": "cloudflare",
+			"start": "next start",
+			"lint": "next lint",
+	```
+
+	Build the app:
+	<PackageManagers type="run" pkg="build:worker" frame="none" />
+
+	Now, you can run Next.js development server:
+	<PackageManagers type="run" pkg="dev" frame="none" />
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: I get an error in the console. I build the worker app and started the Next.js server. Do I need to run the worker as well?

We should encourage the user to open the app in the browser and test it out

**Review by harshil1712 — CHANGES_REQUESTED** [review]
Hey @exectx, this is a great tutorial! I have left a few comments. Will it be possible to simplify the Cursor component? Not a huge issue, but if it can be simplified, it will help the reader follow more easily. Rest looks good! I am excited to see this go live

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx (hunk 34 above)
Comment (exectx) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: it doesn't really matter where, the resulting address will be used in the next step. I make sure the readers leave the worker running for the local setup to work correctly, for instance the local websocket server

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx (hunk 35 above)
Comment (exectx) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: Here I'm updating the `next-rpc/wrangler.toml`, i'll be more explicit

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx (hunk 38 above)
Comment (exectx) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: Did the build fail? I got eslint warnings, mainly caused by the code that highlights incoming/outgoing ws messages, which isn't directly relevant to the main content of the tutorial.

Yes, the durable object worker needs to be running. I'm adding a note

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx (hunk 37 above)
Comment (exectx) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx]: Ok I added more details about how the client side logic works and updated the code to reflect a more realistic scenario. the part responsible for animating the cursor, switching from css transitions to position interpolation.

**Comment by bllchmbrs** [comment]
@exectx , I would recommend adding the label `hyperlint-ignore` to this PR. The indentation isn't something we support well today and this is just too chatty to be useful (IMO). Up to you, but that might help remove some of the extraneous suggestions while we look at how best to support the indentation.

**Comment by exectx** [comment]
@bllchmbrs, thank you for the information but I don't have the necessary permissions for that. Perhaps @harshil1712 or @ToriLindsay could?

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 39]
```diff
@@ -0,0 +1,822 @@
...
+1. Change into your Durable Object Worker directory:
+	```sh
+	cd worker
+	```
+
+	Deploy the Worker:
+	<PackageManagers type="run" pkg="deploy" frame="none" />
+
+	A URL is generated for the Worker, which will be used to update Next.js `NEXT_PUBLIC_WS_HOST` environment variable in your `.env.local` file.
+
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx:804]: Should I copy the whole URL? I copied the whole URL and ran into issues. Let's explicitly tell the reader to only paste the host and not the protocol (`https`)

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 40]
```diff
@@ -0,0 +1,822 @@
...
+	```txt title="next-rpc/.env.local" ins={2} del={1}
+	NEXT_PUBLIC_WS_HOST=localhost:8787
+	NEXT_PUBLIC_WS_HOST=<DEPLOYED-WORKER-HOST>
+	```
+
+2. Change into your root directory and deploy your Next.js app:
```
Comment (harshil1712) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx:810]: Do we still want to use Service bindings in the production? Earlier we told the user we are using it only for local development. So maybe we want to update the steps here?

**Review by harshil1712 — CHANGES_REQUESTED** [review]
Hey @exectx,
Thank you for making the changes. We are at the finish line! I have left two more comments. Once we have resolved them, we should be good to go 🚀

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx (hunk 40 above)
Comment (exectx) [reply src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx:810]: I think it can be an optional step. i'm adding a note mentioning the user can use the durable object rpc method instead of service bindings if they want.

**Review by harshil1712 — APPROVED** [review]
Thank you for making the changes!

**Comment by db-cloudflare** [comment]
Amazing work. @exectx Once your author details are confirmed by email and have been updated on this PR then this should be ready to merge!

File: src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx [hunk 41]
```diff
@@ -0,0 +1,838 @@
+---
+updated: 2024-11-07
+difficulty: Intermediate
+content_type: 📝 Tutorial
+pcx_content_type: tutorial
+title: Build Live Cursors with Next.js, RPC and Durable Objects
+products:
+  - Workers
+  - Durable Objects
```
Comment (db-cloudflare) [inline src/content/docs/workers/tutorials/live-cursors-with-nextjs-rpc-do/index.mdx:9]:
```suggestion
  - Durable Objects
  - Developer Spotlight
spotlight:
  author: Ivan Buendia
  author_bio_link: https://www.linkedin.com/in/ibctx/
  author_bio_source: LinkedIn
```
